### PR TITLE
8261608: Move common CDS archive building code to archiveBuilder.cpp

### PR DESCRIPTION
**Pull request by iklam**
This is a follow-up to https://git.openjdk.java.net/jdk/pull/2296:

- Move common code for writing the CDS archive from metaspaceShared.cpp to archiveBuilder.cpp

- Data structures related to dumping were haphazardly organized in several classes (e.g., `DumpRegions`). We needed various APIs to access them across classes. These should be consolidated in archiveBuilder.cpp and the API should be cleaned up

- Detailed stats (`DumpAllocStats::print_stats`) were available only for static dump. Refactor the code so they are also printed for dynamic dump

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261608](https://bugs.openjdk.java.net/browse/JDK-8261608): Move common CDS archive building code to archiveBuilder.cpp

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 9582e40fbcc313d6adea7d255303dc43feb4865f
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 54e7185fa7587fcc8d1be0ff005fe779bc9ad626

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2536/head:pull/2536`
`$ git checkout pull/2536`

### CODE REVIEW

**Comment by iklam** [comment]
Example of detailed stats that are now available for dynamic dump:

```
[info ][cds] Shared file region (mc )  0:       24 bytes, addr 0x0000000800c30000 file offset 0x00001000 crc 0xa3c1ca20
[info ][cds] Shared file region (rw )  1:     1016 bytes, addr 0x0000000800c31000 file offset 0x00002000 crc 0xb7970a35
[info ][cds] Shared file region (ro )  2:     1632 bytes, addr 0x0000000800c32000 file offset 0x00003000 crc 0x4edeacfa
[info ][cds] Shared file region (bm )  3:      160 bytes, addr 0x0000000000000000 file offset 0x00004000 crc 0x27cf167e
[debug][cds] mc  space:        24 [  0.1% of total] out of      4096 bytes [  0.6% used] at 0x0000000800c30000
[debug][cds] rw  space:      1016 [  6.2% of total] out of      4096 bytes [ 24.8% used] at 0x0000000800c31000
[debug][cds] ro  space:      1632 [ 10.0% of total] out of      4096 bytes [ 39.8% used] at 0x0000000800c32000
[debug][cds] bm  space:       160 [  1.0% of total] out of       160 bytes [100.0% used]
[debug][cds] total    :      2832 [100.0% of total] out of     16384 bytes [ 17.3% used]
[debug][cds] Detailed metadata info (excluding heap regions; rw stats include mc regions):
[debug][cds]                         ro_cnt   ro_bytes     % |   rw_cnt   rw_bytes     % |  all_cnt  all_bytes     %
[debug][cds] --------------------+---------------------------+---------------------------+--------------------------
[debug][cds] Class               :        0          0   0.0 |        1        544  52.3 |        1        544  20.4
[debug][cds] Symbol              :        3         64   3.9 |        0          0   0.0 |        3         64   2.4
[debug][cds] TypeArrayU1         :        3        168  10.3 |        1         40   3.8 |        4        208   7.8
[debug][cds] TypeArrayU2         :        2         16   1.0 |        0          0   0.0 |        2         16   0.6
[debug][cds] TypeArrayU4         :        1         16   1.0 |        0          0   0.0 |        1         16   0.6
[debug][cds] TypeArrayU8         :        2        672  41.2 |        1         40   3.8 |        3        712  26.6
[debug][cds] TypeArrayOther      :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] Method              :        0          0   0.0 |        2        192  18.5 |        2        192   7.2
[debug][cds] ConstMethod         :        2        144   8.8 |        0          0   0.0 |        2        144   5.4
[debug][cds] MethodData          :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] ConstantPool        :        1        312  19.1 |        0          0   0.0 |        1        312  11.7
[debug][cds] ConstantPoolCache   :        0          0   0.0 |        1        136  13.1 |        1        136   5.1
[debug][cds] Annotations         :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] MethodCounters      :        0          0   0.0 |        1         64   6.2 |        1         64   2.4
[debug][cds] RecordComponent     :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] SymbolHashentry     :        3         32   2.0 |        0          0   0.0 |        3         32   1.2
[debug][cds] SymbolBucket        :        2         16   1.0 |        0          0   0.0 |        2         16   0.6
[debug][cds] StringHashentry     :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] StringBucket        :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] ModulesNatives      :        0          0   0.0 |        0          0   0.0 |        0          0   0.0
[debug][cds] Other               :        0        192  11.8 |        0         24   2.3 |        0        216   8.1
[debug][cds] --------------------+---------------------------+---------------------------+--------------------------
[debug][cds] Total               :       19       1632 100.0 |        7       1040 100.0 |       26       2672 100.0
```

File: src/hotspot/share/classfile/moduleEntry.cpp [hunk 1]
```diff
@@ -33,7 +33,6 @@
 #include "memory/archiveUtils.hpp"
```
Comment (coleenp) [inline src/hotspot/share/classfile/moduleEntry.cpp:33]: Do these files need archiveUtils.hpp or does archiveBuilder.hpp need it?

File: src/hotspot/share/memory/archiveBuilder.cpp [hunk 2]
```diff
@@ -171,11 +176,8 @@ ArchiveBuilder::ArchiveBuilder(DumpRegion* mc_region, DumpRegion* rw_region, Dum
   _num_instance_klasses = 0;
   _num_obj_array_klasses = 0;
   _num_type_array_klasses = 0;
```
Comment (coleenp) [inline src/hotspot/share/memory/archiveBuilder.cpp:178]: Should these be initializers also?

File: src/hotspot/share/memory/archiveUtils.hpp [hunk 3]
```diff
@@ -39,15 +40,18 @@ class VirtualSpace;
 // mark_pointer(/*ptr_loc=*/&k->_name). It's required that (_prt_base <= ptr_loc < _ptr_end). _ptr_base is
 // fixed, but _ptr_end can be expanded as more objects are dumped.
 class ArchivePtrMarker : AllStatic {
-  static CHeapBitMap* _ptrmap;
-  static address*     _ptr_base;
-  static address*     _ptr_end;
+  static CHeapBitMap*  _ptrmap;
+  static VirtualSpace* _vs;
```
Comment (coleenp) [inline src/hotspot/share/memory/archiveUtils.hpp:44]: I think this is a good change.

File: src/hotspot/share/memory/metaspaceShared.inline.hpp [hunk 4]
```diff
@@ -1,39 +0,0 @@
...
-#ifndef SHARE_MEMORY_METASPACESHARED_INLINE_HPP
-#define SHARE_MEMORY_METASPACESHARED_INLINE_HPP
-
-#include "memory/metaspaceShared.hpp"
-#if INCLUDE_G1GC
-#include "gc/g1/g1Allocator.inline.hpp"
-#endif
-
-#if INCLUDE_CDS_JAVA_HEAP
-bool MetaspaceShared::is_archive_object(oop p) {
```
Comment (coleenp) [inline src/hotspot/share/memory/metaspaceShared.inline.hpp:34]: Was this unused? I don't find 'is_archive_object' in the diff.

**Review by coleenp — APPROVED** [review]
Looks like a good change.

File: src/hotspot/share/memory/metaspaceShared.inline.hpp (hunk 4 above)
Comment (iklam) [reply src/hotspot/share/memory/metaspaceShared.inline.hpp:34]: This whole file was unused (and the function was not even declared in the MetaspaceShared class) so I removed it.

File: src/hotspot/share/memory/archiveBuilder.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/memory/archiveBuilder.cpp]: I changed them to initializers.

File: src/hotspot/share/classfile/moduleEntry.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/moduleEntry.cpp:33]: Both packageEntry.cpp and moduleEntry.cpp use ArchivePtrMarker which is defined in archiveUtils.hpp.

**Review by coleenp — APPROVED** [review]
Looks good!

File: src/hotspot/share/memory/archiveBuilder.cpp [hunk 5]
```diff
@@ -159,37 +161,40 @@ void ArchiveBuilder::SourceObjList::relocate(int i, ArchiveBuilder* builder) {
...
-
-  _requested_static_archive_bottom = NULL;
-  _requested_static_archive_top = NULL;
-  _mapped_static_archive_bottom = NULL;
-  _mapped_static_archive_top = NULL;
-  _requested_dynamic_archive_bottom = NULL;
-  _requested_dynamic_archive_top = NULL;
-  _buffer_to_requested_delta = 0;
+  assert(_current == NULL, "must be");
+  _current = this;
```
Comment (calvinccheung) [inline src/hotspot/share/memory/archiveBuilder.cpp:197]: These lines used to be at the beginning of the function. Any reasons why they are moved?

File: test/hotspot/jtreg/runtime/cds/appcds/LotsOfClasses.java [hunk 6]
```diff
@@ -49,7 +49,8 @@ public static void main(String[] args) throws Exception {
         opts.addSuffix("-Xlog:hashtables");
         opts.addSuffix("-Xmx500m");
         opts.addSuffix("-Xlog:gc+region+cds");
-        opts.addSuffix("-Xlog:gc+region=trace");
+        //opts.addSuffix("-Xlog:gc+region=trace");
+        opts.addSuffix("-Xlog:cds=debug");  // test detailed metadata info printing
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/LotsOfClasses.java:53]: Remove the commented line #52?

**Review by calvinccheung — COMMENTED** [review]
Looks good overall. Couple of minor comments.

File: src/hotspot/share/memory/archiveBuilder.cpp (hunk 5 above)
Comment (iklam) [reply src/hotspot/share/memory/archiveBuilder.cpp:197]: I think it's better to set `_current` after the instance has been fully initialized.

File: test/hotspot/jtreg/runtime/cds/appcds/LotsOfClasses.java (hunk 6 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/LotsOfClasses.java]: I removed it.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @coleenp and @calvinccheung for your review!
/integrate

**Pull request closed without merging**